pkg/registry/flowcontrol/ensurer/strategy.go

```diff
@@ -0,0 +1,328 @@
+	CopySpec(bootstrap, current runtime.Object) error
+}
+
+// this internal interface provides abstraction for CRUD operation
+// related to both 'FlowSchema' and 'PriorityLevelConfiguration' objects.
+// Since the ensure logic for both types is common, we use a few internal interfaces
+// to abstract out the differences of these two types.
+type configurationClient interface {
+	Create(object runtime.Object) (runtime.Object, error)
+	Update(object runtime.Object) (runtime.Object, error)
+	Get(name string) (configurationObject, error)
+	Delete(name string) error
+}
+
+type configurationWrapper interface {
+	// TypeName returns the type of the configuration that this interface deals with.
+	// We use it to log the type name of the configuration object being ensured.
+	// It is either 'PriorityLevelConfiguration' or 'FlowSchema'
+	TypeName() string
+
+	configurationClient
+	specCopier
+}
+
+// A convenient wrapper interface that is used by the ensure logic.
+type configurationObject interface {
+	metav1.Object
+	runtime.Object
+}
+
+// auto-update strategy for the suggested configurationWrapper objects
+type suggestedEnsureStrategy struct {
+	copier       specCopier
+	shouldCreate bool
+}
+
+func (s *suggestedEnsureStrategy) Name() string {
+	return "suggested"
+}
+
+func (s *suggestedEnsureStrategy) ShouldCreate() bool {
+	return s.shouldCreate
+}
+
+func (s *suggestedEnsureStrategy) ShouldUpdate(current, bootstrap configurationObject) (runtime.Object, bool, error) {
+	if current == nil || bootstrap == nil {
+		return nil, false, nil
+	}
+
+	autoUpdateSpec := shouldUpdateSpec(current)
+	updateAnnotation := shouldUpdateAnnotation(current, autoUpdateSpec)
+
+	var specChanged bool
+	if autoUpdateSpec {
+		changed, err := s.copier.HasSpecChanged(bootstrap, current)
+		if err != nil {
+			return nil, false, fmt.Errorf("failed to compare spec - %w", err)
+		}
+		specChanged = changed
+	}
+
+	if !(updateAnnotation || specChanged) {
+		// the annotation key is up to date and the spec has not changed, no update is necessary
+		return nil, false, nil
+	}
+
+	// if we are here, either we need to update the annotation key or the spec.
+	copy, ok := current.DeepCopyObject().(configurationObject)
+	if !ok {
+		// we should never be here
+		return nil, false, errors.New("incompatible object type")
+	}
+
+	if updateAnnotation {
+		setAutoUpdateAnnotation(copy, autoUpdateSpec)
+	}
+	if specChanged {
+		s.copier.CopySpec(bootstrap, copy)
+	}
+
+	return copy, true, nil
+}
+
+// auto-update strategy for the mandatory configurationWrapper objects
+type mandatoryEnsureStrategy struct {
+	copier specCopier
+}
+
+func (s *mandatoryEnsureStrategy) Name() string {
+	return "mandatory"
+}
+
+func (s *mandatoryEnsureStrategy) ShouldCreate() bool {
+	return true
+}
+
+func (s *mandatoryEnsureStrategy) ShouldUpdate(current, bootstrap configurationObject) (runtime.Object, bool, error) {
+	if current == nil || bootstrap == nil {
```

Comment:
There is still significant code duplication for this method before mandatory and suggest.

The only difference is that for mandatory "shouldUpdateSpec" should always return true.
We should be able to simplify a lot by just parametrizing this function.

Reply:
yeah, i have one `strategy` implementation now but it varies with `alwaysAutoUpdateSpec` which is true for manadatory strategy.